### PR TITLE
Replicas: Improve sorting replicas; #4017

### LINKED ISSUE #4017: Set geoip as default sorting algorithm in list_replicas
Motivation
----------
Currently list_replicas is returning the list of replicas in semi-random order. It does return `lan` first and then `wan`, however the `wan` replicas are not sorted. Specifically if only `wan` replicas are available this is an issue.

Modification
------------
The default sorting should be changed to geoip.

### PR DESCRIPTION
**Pull request by bziemons**
Improve sorting replicas

Deduplicate code related to replica sorting
Set geoip as default sorting option for all
Return lan locations always before wan locations for list_replicas
Replace tests for geoip replica sorting
Fix setting priority to the order of what the sort returns on list_replicas endpoint

### CODE REVIEW

**Comment by bziemons** [comment]
Depends on #4081

File: lib/rucio/web/rest/webpy/v1/redirect.py [hunk 1]
```diff
@@ -142,17 +143,7 @@ def GET(self, scope, name):
...
-                    replicas = sort_geoip(dictreplica, client_location['ip'], ignore_error=True)
-                elif select == 'closeness':
-                    replicas = sort_closeness(dictreplica, client_location)
-                elif select == 'dynamic':
-                    replicas = sort_dynamic(dictreplica, client_location)
-                elif select == 'ranking':
-                    replicas = sort_ranking(dictreplica, client_location)
-                else:
-                    replicas = sort_random(dictreplica)
+                replicas = sort_replicas(dictreplica, client_location, selection=select, default='random')
```
Comment (mlassnig) [inline lib/rucio/web/rest/webpy/v1/redirect.py]: should this be default=geoip instead?

File: lib/rucio/web/rest/webpy/v1/redirect.py [hunk 2]
```diff
@@ -277,17 +268,7 @@ def GET(self, scope, name):
...
-                            elif select == 'closeness':
-                                rep = sort_closeness(dictreplica, client_location)
-                            elif select == 'dynamic':
-                                rep = sort_dynamic(dictreplica, client_location)
-                            elif select == 'ranking':
-                                rep = sort_ranking(dictreplica, client_location)
-                            else:
-                                rep = sort_random(dictreplica)
-
+                            rep = sort_replicas(dictreplica, client_location, selection=select, default='random')
```
Comment (mlassnig) [inline lib/rucio/web/rest/webpy/v1/redirect.py]: same here

File: lib/rucio/web/rest/flaskapi/v1/redirect.py [hunk 3]
```diff
@@ -141,17 +142,7 @@ def generate():
...
-                        replicas = sort_geoip(dictreplica, client_location['ip'], ignore_error=True)
-                    elif select == 'closeness':
-                        replicas = sort_closeness(dictreplica, client_location)
-                    elif select == 'dynamic':
-                        replicas = sort_dynamic(dictreplica, client_location)
-                    elif select == 'ranking':
-                        replicas = sort_ranking(dictreplica, client_location)
-                    else:
-                        replicas = sort_random(dictreplica)
+                    replicas = sort_replicas(dictreplica, client_location, selection=sortby, default='random')
```
Comment (mlassnig) [inline lib/rucio/web/rest/flaskapi/v1/redirect.py]: default=geoip?

File: lib/rucio/web/rest/flaskapi/v1/redirect.py [hunk 4]
```diff
@@ -283,17 +274,7 @@ def get(self, scope_name):
...
-                            elif select == 'closeness':
-                                rep = sort_closeness(dictreplica, client_location)
-                            elif select == 'dynamic':
-                                rep = sort_dynamic(dictreplica, client_location)
-                            elif select == 'ranking':
-                                rep = sort_ranking(dictreplica, client_location)
-                            else:
-                                rep = sort_random(dictreplica)
-
+                            rep = sort_replicas(dictreplica, client_location, selection=select, default='random')
```
Comment (mlassnig) [inline lib/rucio/web/rest/flaskapi/v1/redirect.py]: same here?

File: lib/rucio/web/rest/webpy/v1/redirect.py (hunk 1 above)
Comment (bziemons) [reply lib/rucio/web/rest/webpy/v1/redirect.py]: no, the redirect endpoint is out of scope.

File: lib/rucio/web/rest/webpy/v1/redirect.py (hunk 2 above)
Comment (bziemons) [reply lib/rucio/web/rest/webpy/v1/redirect.py]: no, the redirect endpoint is out of scope.

File: lib/rucio/web/rest/flaskapi/v1/redirect.py (hunk 3 above)
Comment (bziemons) [reply lib/rucio/web/rest/flaskapi/v1/redirect.py]: no, the redirect endpoint is out of scope.

File: lib/rucio/web/rest/flaskapi/v1/redirect.py (hunk 4 above)
Comment (bziemons) [reply lib/rucio/web/rest/flaskapi/v1/redirect.py]: no, the redirect endpoint is out of scope.

**Review by bari12 — APPROVED** [review]

**Pull request merged**